File: programers_Lv2/[3차] 압축.py
# https://school.programmers.co.kr/learn/courses/30/lessons/17684
import string

def solution(msg):
    answer = []
    new_value = 27
    alpa_dict = {v: i+1 for i, v in enumerate(string.ascii_uppercase)}

    new_msg = msg
    temp_str = ''
    while True:
        try:
            temp_str += new_msg[0]
            if temp_str in alpa_dict:
                new_msg = new_msg[1:]
                continue
            else:
                alpa_dict[temp_str] = new_value
                new_value += 1
                answer.append(alpa_dict[temp_str[:-1]])
                temp_str = ''
        except:
            answer.append(alpa_dict[temp_str])
            return answer
        
print(solution("KAKAO"))  # [11, 1, 27, 15]

File: programers_Lv2/[3차] 압축.js
function solution(msg) {
  var answer = [];
  let newNum = 27;
  const alpaStr = new Array(26)
    .fill()
    .map((_, i) => String.fromCharCode(i + 65))
    .join("");
  const alpaObj = alpaStr.split("").reduce((acc, curr, idx) => {
    return { ...acc, [curr]: idx + 1 };
  }, {});

  let newMsg = msg;
  let tempStr = "";

  while (newMsg.length > 0) {
    tempStr += newMsg.slice(0, 1);
    if (alpaObj[tempStr] !== undefined) {
      newMsg = newMsg.slice(1, newMsg.length);
      continue;
    } else {
      alpaObj[tempStr] = newNum;
      newNum++;
      answer.push(alpaObj[tempStr.slice(0, -1)]);
      tempStr = "";
    }
  }
  answer.push(alpaObj[tempStr]);
  return answer;
}

console.log(solution("KAKAO")); // [11, 1, 27, 15]
